Reduce ZF request tail latency improve stability with concurrent hedging and circuit breaker

CodeBoy2006 opened the pull request:
## 背景

目前 wejh 通过 wejh-go 调用 funnel，再由 funnel 访问正方教务（ZF / OAuth）接口。

线上已知问题包括：

1. ZF 接口延迟大且不稳定，当前逻辑为 "单条线路串行重试" ，导致接口 p99 较高；
2. 启动后所有线路状态初始化为健康，小流量时难以在时间窗口内累计足够失败次数，导致 "线路已经不稳定，但仍被当做健康线路" ；
3. 内外网 + ZF/OAuth 多条线路质量差异大，目前仅做随机负载均衡，无法充分利用延迟更优的线路；
4. 夜间仅外网 ZF 可用，其它线路在固定时间段集体失效，断路阈值固定为 50，窗口期较长；
5. 可观测性较差，无法直观区分请求实际走了哪条线路，排查超时问题成本较高。

在现阶段的实际运行环境中：

- wejh-go 所在机器的 IP 已加入 ZF 白名单；
- 实际 QPM 极低，对 ZF 端的额外请求负载非常有限。

在这样的前提下，采用 "无脑并发对冲" 的策略，可以用较小的代价大幅降低尾延迟并提高成功率。

---

## 效果预测

经过模拟分析，随着并发量的提升，接口的所有指标都有极为显著的改善。因此，此方案也能大幅压缩 平均、p95、p99 延迟并提升请求的成功率：

<img width="3085" height="1646" alt="image" src="https://github.com/user-attachments/assets/77dbaa77-d78a-40c5-8a40-946b1df231ee" />

*（橙色为WeJH-Go的原逻辑，绿色为 SWRR 负载均衡 + 2 并发对冲，蓝色为本方案实现的全量对冲）*

---

## 本 PR 的主要改动

### 1. funnel 侧（ `app/services/funnelServices/funnelServices.go`）：引入 ZF 接口的并发对冲

使得 ZF 场景下的调用变为并发请求对冲，在小 QPM 前提下，可以显著压缩 p99 延迟并提升成功率。

### 2. circuitBreaker 侧（`app/utils/circuitBreaker/loadbalance.go`）：暴露可用节点列表

---

## 设计思路简述

1. **面向现有问题而不是推倒重来：**   
   在不重构整体负载均衡架构、不引入额外微服务框架和云原生组件的前提下，先解决影响用户体验的 ZF 接口 p99 延迟问题。

2. **充分利用现有熔断与健康检测：**   
   当前项目已经有 `circuitBreaker` 模块和定期健康检查任务，本次改动不重复造轮子，只是让 "对冲请求" 也参与到失败/成功统计中，从而推动线路熔断与恢复。
  由于全量并发对冲的特殊性，因此不需要负载均衡的机制。

3. **利用白名单 + 小 QPM 的现实条件：**   
   相比复杂的「智能路由 + 动态权重」，当前阶段在 wejh-go 侧做并发对冲，能以更少的工程复杂度获取更明显的收益，对 ZF 端压力也可接受。

4. **为后续迭代保留空间：**   
   新增的 `LoadBalance.List` 与对冲逻辑都是中立的：
   - 未来可以在 LB 层引入权重、Redis 持久化、时间段配置；
   - 也可以把 LB 的实现迁移到其他框架，而保持 funnel 对 `List` / `Pick` 的依赖不变。

---

## 影响

- **对调用方兼容：** 
  - `FetchHandleOfPost` 函数签名和返回语义保持不变；
  - 非 ZF 接口仍旧走单节点逻辑，不受影响；
  - controller / service 层无需改动。

- **对 ZF 服务端的影响：** 
  - 单次请求可能会并发打向多条 ZF 线路，使得本服务流向正方服务器的流量增加；
  - 在当前 QPM 较低且 IP 已在 ZF 白名单内的前提下，此方案具有可行性；
  - 若未来流量显著增长，可限制参与对冲的线路数量，或按权重裁剪候选列表。

- **对熔断和健康检测的影响：** 
  - 各线路的失败/成功统计更加充分，熔断判断更敏感；
  - 小流量场景下，即使不能快速熔断，用户也能通过对冲获得更高的成功率（特定线路的失败对用户总体无感）。

---

## 小结

本 PR 集中解决了 ZF 接口高 p99 延迟和不稳定的问题，通过简单有效的 **全量并发对冲** 策略大幅提升了用户侧体验。

请审查并给出进一步建议 🙌

CodeBoy2006 committed:
feat(funnel): 并发对冲正方请求并接入现有熔断逻辑

(cherry picked from commit 57010c1becd576fc26ee7f634c4a947bcda3cf53)

Penryn commented on app/services/funnelServices/funnelServices.go:
```diff
@@ -1,70 +1,198 @@
+		for _, h := range hosts {
+			if h == host {
+				found = true
+				break
```
已存在好像没做处理是吧

CodeBoy2006 replied:
对，这里之前只做了“避免重复追加”，确实没有把传进来的 host 提升优先级。
已修改

Penryn commented on app/services/funnelServices/funnelServices.go:
```diff
@@ -1,70 +1,198 @@
+			default:
+			}
+
+			rc, err := singleHostRequest(h, api, form)
```
是不是可以考虑让singleHostRequest也接收 context ，已发出的请求也可以取消

Penryn replied:
目前好像仅对发送请求时做了拦截

CodeBoy2006 replied:
有道理，已经把 `singleHostRequest` 改成接收 `ctx`

CodeBoy2006 committed:
refactor(funnel): improve host prioritization & add context-aware request cancellation

cbluebird commented on app/services/funnelServices/funnelServices.go:
```diff
@@ -1,70 +1,211 @@
+		go func() {
+			defer wg.Done()
+
+			// 如果已经有其他节点成功了，尽量避免无意义请求
```
需要recover，独立的goroutine 不能被gin 的recover 包住

CodeBoy2006 replied:
已加

cbluebird commented on app/services/funnelServices/funnelServices.go:
```diff
@@ -1,70 +1,211 @@
+	}
+
+	// 等所有协程结束后关闭通道
+	go func() {
```
defer 不行吗

CodeBoy2006 replied:
已加

cbluebird commented on app/services/funnelServices/funnelServices.go:
```diff
@@ -1,70 +1,211 @@
+		}
+
+		switch r.rc.Code {
+		case 200:
```
搞个枚举，带上注释

CodeBoy2006 replied:
已从 funnel 端同步状态码枚举

cbluebird commented:
ai 写的总结吧，搞这么多

CodeBoy2006 commented on app/services/funnelServices/funnelServices.go:
```diff
@@ -1,70 +1,211 @@
+// - 非 ZF 接口：单节点调用（保留原逻辑）
+// - ZF 接口：并发对冲到所有当前可用节点 + 简单熔断 / 恢复
+func FetchHandleOfPost(form url.Values, host string, api funnelApi.FunnelApi) (interface{}, error) {
 	loginType := funnelApi.LoginType(form.Get("type"))
```
和业务逻辑不符，不采纳

CodeBoy2006 commented on app/services/funnelServices/funnelServices.go:
```diff
@@ -1,70 +1,211 @@
+	// 竞争结果：
+	// - 第一个 200 直接返回数据，并标记该节点 Success
+	// - 第一个 412 / 416 属于用户态错误，也直接返回，不再等其它节点
+	// - 413 视为该节点过载/异常，Fail 一次，继续等待其它节点
```
413 是 funnel 的验证码错误

CodeBoy2006 replied:
验证码错误不应计入fail，但是不知道目前 funnel 这块的逻辑有无发生变动

CodeBoy2006 commented on app/services/funnelServices/funnelServices.go:
```diff
@@ -1,70 +1,211 @@
+			hosts = append([]string{host}, hosts...)
+		} else if idx > 0 {
+			// 已存在：挪到最前面，避免重复
+			hosts = append([]string{host}, append(hosts[:idx], hosts[idx+1:]...)...)
```
合理

CodeBoy2006 replied:
之前的写法太唐了

CodeBoy2006 commented on app/services/funnelServices/funnelServices.go:
```diff
@@ -1,70 +1,211 @@
+	// - 413 视为该节点过载/异常，Fail 一次，继续等待其它节点
+	for r := range resultCh {
+		if r.err != nil {
+			// 网络 / 解析错误 / ctx 取消：认为节点异常
```
这个完全正确

CodeBoy2006 commented on app/services/funnelServices/funnelServices.go:
```diff
@@ -1,70 +1,211 @@
+				return
+			default:
+			}
+
```
收益不大，容易让代码变脏，不采纳。

CodeBoy2006 committed:
refactor(funnel): address review feedback for ZF hedging & load balancer

已经合并：
- goroutine recover
- ctx 取消不再记 Fail
- host 优先级 swap
- 从 funnel 端同步了状态码枚举常量，带注释
- range 闭包用传参方式
- close(resultCh) 用 defer

CodeBoy2006 commented on app/services/funnelServices/funnelServices.go:
```diff
@@ -63,21 +67,27 @@ func FetchHandleOfPost(form url.Values, host string, api funnelApi.FunnelApi) (i
 			return nil, apiException.OAuthNotUpdate
+
+		// 410 / 413 / 414 / 415 以及其它未知 code：
+		// 统一视为服务侧异常，向上抛 ServerError
```
funnel 端如果已经下沉了验证码重试的逻辑，那么 WeJH-Go 端去除重试流程是合理的。
如果还没有，则暂时需要保留原有的重试逻辑。

CodeBoy2006 replied:
已加回重试流程，保证兼容性

CodeBoy2006 committed:
rollback: add back the retry mechanism for incorrect captcha entries

CodeBoy2006 commented on app/utils/circuitBreaker/loadbalance.go:
```diff
@@ -78,14 +102,31 @@ func (b *randomLB) LoadBalance() LoadBalanceType {
+		return nil
+	}
+	out := make([]string, b.Size)
+	copy(out, b.Api)
```
合理

cbluebird commented on app/utils/circuitBreaker/loadbalance.go:
```diff
@@ -39,6 +43,25 @@ func (lb *LoadBalance) Pick(zfFlag, oauthFlag bool) (string, funnelApi.LoginType
+// List 返回当前可用后端节点的快照
+// - loginType 为 Oauth：返回 OAuth 池
+// - 其它（ZF / Unknown）：统一返回 ZF 池
+func (lb *LoadBalance) List(loginType funnelApi.LoginType) []string {
```
优化一下，未知的返回 error，并打印日志

CodeBoy2006 committed:
fix: return error for unknown load balancer login type

Penryn committed:
Merge branch 'dev' into feat/hedge-zf
